taichi/backends/vulkan/vulkan_device.cpp

```diff
@@ -0,0 +1,602 @@
+    alloc_info.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT;
+    alloc_info.preferredFlags = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT |
+                                VK_MEMORY_PROPERTY_HOST_CACHED_BIT;
+  } else if (params.host_read) {
+    alloc_info.usage = VMA_MEMORY_USAGE_GPU_TO_CPU;
+  } else if (params.host_write) {
+    alloc_info.usage = VMA_MEMORY_USAGE_CPU_TO_GPU;
+  } else {
+    alloc_info.usage = VMA_MEMORY_USAGE_GPU_ONLY;
+  }
+
+  vmaCreateBuffer(allocator_, &buffer_info, &alloc_info, &alloc.buffer,
+                  &alloc.allocation, &alloc.alloc_info);
+
+  return handle;
+}
+
+void VulkanDevice::dealloc_memory(DeviceAllocation allocation) {
+  AllocationInternal &alloc = allocations_[allocation.alloc_id];
+
+  vmaDestroyBuffer(allocator_, alloc.buffer, alloc.allocation);
+
+  allocations_.erase(allocation.alloc_id);
+}
+
+void *VulkanDevice::map_range(DevicePtr ptr, uint64_t size) {
+  AllocationInternal &alloc_int = allocations_.at(ptr.alloc_id);
+
+  TI_ASSERT_INFO(alloc_int.mapped == nullptr,
+                 "Memory can not be mapped multiple times");
+
+  vkMapMemory(device_, alloc_int.alloc_info.deviceMemory,
+              alloc_int.alloc_info.offset + ptr.offset, size, 0,
+              &alloc_int.mapped);
+
+  return alloc_int.mapped;
+}
+
+void *VulkanDevice::map(DeviceAllocation alloc) {
+  AllocationInternal &alloc_int = allocations_.at(alloc.alloc_id);
+
+  TI_ASSERT_INFO(alloc_int.mapped == nullptr,
+                 "Memory can not be mapped multiple times");
+
+  vkMapMemory(device_, alloc_int.alloc_info.deviceMemory,
+              alloc_int.alloc_info.offset, alloc_int.alloc_info.size, 0,
+              &alloc_int.mapped);
+
+  return alloc_int.mapped;
+}
+
+void VulkanDevice::unmap(DevicePtr ptr) {
+  AllocationInternal &alloc_int = allocations_.at(ptr.alloc_id);
+
+  TI_ASSERT_INFO(alloc_int.mapped, "Memory is not mapped");
+
+  vkUnmapMemory(device_, alloc_int.alloc_info.deviceMemory);
+  alloc_int.mapped = nullptr;
+}
+
+void VulkanDevice::unmap(DeviceAllocation alloc) {
+  AllocationInternal &alloc_int = allocations_.at(alloc.alloc_id);
+
+  TI_ASSERT_INFO(alloc_int.mapped, "Memory is not mapped");
+
+  vkUnmapMemory(device_, alloc_int.alloc_info.deviceMemory);
+  alloc_int.mapped = nullptr;
+}
+
+void VulkanDevice::memcpy_internal(DevicePtr dst,
+                                   DevicePtr src,
+                                   uint64_t size) {
+}
+
+std::unique_ptr<CommandList> VulkanDevice::new_command_list() {
+  VkCommandBuffer buffer = VK_NULL_HANDLE;
+
+  if (free_cmdbuffers_.size()) {
+    buffer = free_cmdbuffers_.back();
+    free_cmdbuffers_.pop_back();
+  } else {
+    // FIXME: Proper multi queue support, instead of defaulting to compute
+    // command pool
+    VkCommandBufferAllocateInfo alloc_info{};
+    alloc_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+    alloc_info.commandPool = compute_cmd_pool();
```

Comment:
Can we add an argument to `new_command_list` (e.g. a `struct CommandListParams`) so that we can optionally use the graphics cmd pool?

Reply:
Sounds about right. Should we also assume all graphics queue will carry compute capability? Benefit of that being not bothering with cross-queue synchronization.

Reply:
Hmm. I would slightly prefer not making that assumption. At some point we will need to properly handle graphics-compute synchronization anyways, since the compute backend isn't always Vulkan. And at that point it might be easier to use two queues even though both backends are vulkan.

Reply:
If the compute backend isn't Vulkan the compute should be on another different Device, maybe a CudaDevice class. Data sharing through those should somehow be achieved with the `share_to()` call. I think for all the APIs I can think of, unless it's old OpenGL, a graphics queue always contain compute capabilities, but a lot of compute queues are compute only

Reply:
If the compute backend and graphics backend are both Vulkan, will we be using the same `Device` for both? It might be the case that we will have cleaner code if we always created two devices, one for compute and one for graphics. This way,  we can re-use some of the logic for compute-graphics data sharing for different compute backends.

Reply:
Using two would be creating an assumption that the memory between them are not unified, which could be a potential performance hurdle.

Reply:
Ideally in the future taichi will create a device per detected physical device. So a user with a Nvidia card will boot up taichi with both a VulkanDevice and a CudaDevice. If the user has another integrated intel, we might also have another VulkanDevice maps to that intel GPU

Reply:
Left as a TODO